### PR TITLE
Ethashproof Patch

### PR DESCRIPTION
**Pull request by AurevoirXavier**

### CODE REVIEW

File: frame/bridge/eth/relay/src/lib.rs [hunk 1]
```diff
@@ -253,50 +254,31 @@ decl_module! {
 		/// - Up to one event
 		/// # </weight>
 		#[weight = SimpleDispatchInfo::FixedNormal(200_000)]
-		pub fn relay_header(origin, header: EthHeader) {
+		pub fn relay_header(origin, header: EthHeader, ethash_proof: Vec<DoubleNodeWithMerkleProof>) {
```
[the remark this answers is not in the document]
Comment (HackFisher) [reply frame/bridge/eth/relay/src/lib.rs:257]: ethash_proof is an optional parameter here?

File: frame/bridge/eth/relay/src/lib.rs [hunk 2]
```diff
@@ -123,31 +119,44 @@ decl_storage! {
...
 
-		pub CheckAuthorities get(fn check_authorities) config(): bool = true;
+		pub CheckAuthority get(fn check_authority) config(): bool = true;
 		pub Authorities get(fn authorities) config(): Vec<T::AccountId>;
 	}
 	add_extra_genesis {
 		// genesis: Option<Header, Difficulty>
 		config(genesis_header): Option<(u64, Vec<u8>)>;
-		config(dag_merkle_roots): Vec<H128>;
+		config(dag_merkle_roots): DagMerkleRoots;
```
Comment (HackFisher) [inline frame/bridge/eth/relay/src/lib.rs:137]: DagMerkleRoots is defined in std, can be used here?

File: frame/bridge/eth/relay/src/lib.rs [hunk 3]
```diff
@@ -565,9 +527,66 @@ impl<T: Trait> Module<T> {
...
+
+		let pair = ethash::hashimoto(
+			header_hash.clone(),
+			nonce.clone(),
+			ethash::get_full_size(block_number as usize / 30000),
+			|offset| {
+				let index = index.replace_with(|&mut old| old + 1);
+
+				// Each two nodes are packed into single 128 bytes with Merkle proof
+				let node = if let Some(node) = nodes.get(index / 2) {
```
[the remark this answers is not in the document]
Comment (AurevoirXavier) [reply frame/bridge/eth/relay/src/lib.rs:551]: get `None` means, there's nothing at this index

File: frame/bridge/eth/relay/src/lib.rs [hunk 4]
```diff
@@ -253,50 +254,31 @@ decl_module! {
...
-				ensure!(Self::authorities().contains(&relayer), <Error<T>>::AccountNP);
+			// 1. proof of difficulty
+			// 2. proof of pow (mixhash)
+			// 3. challenge
+			{
+				Self::verify_header_basic(&header)?;
+				// FIXME:
+				// waiting offchain-tool's update
+				// pass empty `ethash_proof` to skip this verify
+				if !ethash_proof.is_empty() {
```
[the remark this answers is not in the document]
Comment (AurevoirXavier) [reply frame/bridge/eth/relay/src/lib.rs:277]: this is the final api, offchain tools should be upgrade

File: frame/bridge/eth/relay/src/lib.rs (hunk 2 above)
Comment (HackFisher) [reply frame/bridge/eth/relay/src/lib.rs:137]: https://substrate.dev/docs/en/1.0/runtime/types/genesisconfig-struct#setting-the-genesis-configuration-values

**Pull request merged**